### ISSUE TITLE
Tiered expert residency: static hot/cold expert slab replacing the demand-staged MoE cache

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/llama.cpp#132]
# Tiered expert residency: a static hot/cold expert slab, replacing the demand-staged MoE cache

## Goal

Keep frequently-used experts permanently in VRAM and compute the rest on the CPU, with
**no per-dispatch transfers**. Target: **29-31 t/s** on the RTX 3060 against the current
best of **21.3 t/s**, at equal VRAM.

This is not a new cache policy. It removes the cache from the decode path entirely and
replaces it with a residency decision made once, at load time.

## Why the current MoE cache cannot be tuned into this

The cache answers *"is this expert resident?"* at runtime, per route, 1440 times per
dispatch, and answers a miss by **staging the expert over PCIe**. On this rig that is the
most expensive of the three available paths:

| path | ms per expert (2.148 MiB) | ratio |
| --- | --- | --- |
| already in VRAM | **0.0029** | 1x |
| compute on CPU (host DDR) | **0.0677** | 23x |
| stage over PCIe x4 | **0.342** | **118x** |

Break-even against `cache-0 + --cpu-moe` (arm 000, 130.8 ms/dispatch):

```
affordable link time = 97.5 ms   (what the CPU does the same work for)
cost per miss        = 0.342 ms x 48 layers
=> affordable misses = 5.94 per layer of 30 routes
=> REQUIRED HIT RATE > 80%
```

The measured skew ceiling (exp1 log: `top1_pct=12.72 top5_pct=42.70 top10_pct=64.18`,
read as top-N% of experts) puts a **perfect oracle** at N=42 slots — 8.2% of 512 experts —
at roughly **57-64%**.

> Break-even needs >80%. The ceiling is ~64%. **No policy, prior, or prefetch can close
> that gap at any N this box can hold.** This is why every tuning lever in #129/#130 came
> back dead: they were all optimising a quantity whose maximum is below break-even.

Five arms confirmed this empirically without the split ever engaging once (see #130).

## Design

Residency becomes a **property of the expert's id**, fixed at load time. `id < N_hot`
means VRAM. It is a comparison, not a lookup — so there is no plan to build, no
reader-row mapping to reconstruct, and no admission ordering to get right. Those are
precisely the three things that blocked the previous attempt.

```
LOAD TIME (once)
  rank experts per layer by frequency (offline trace tooling exists)
  permute ffn_gate_inp output rows by that ranking
      -> the router now EMITS tier-ordered ids; no runtime remap table
  split each layer's permuted expert tensor:
      HOT  [n_embd, n_ff, N_hot + 1]   -> VRAM     (+1 = null expert, zeros)
      COLD [n_embd, n_ff, 512-N_hot+1] -> host RAM (+1 = null expert, zeros)

RUN TIME (per layer)
  ids --+--> ids_hot  = (id <  N_hot) ? id         : N_hot
        +--> ids_cold = (id >= N_hot) ? id - N_hot : 512-N_hot
             (one small CUDA kernel over a ~120-byte i32 tensor)

  mul_mat_id(HOT,  ids_hot )   on CUDA
  mul_mat_id(COLD, ids_cold)   on CPU
                |
                +--> add partials
```

Summing partials is **exact**, not an approximation: MoE output is `sum_i w_i * E_i(x)`.

### The null expert

Both passes run at full `n_expert_used` width, so there is no ragged batching, no
compaction, and **no new ggml op is required**. Slots that do not belong to a pass point
at a zero-filled null expert and contribute nothing.

On the CPU that null is read once per layer and stays in L3, so the unused slots cost
about one expert instead of ten. Measured tax: **3.25 ms/dispatch (3.6%)**. Removing it
via id compaction is a possible v2; it is not needed for the headline result.

## Keeping it adaptive

Promote a cold expert into the hot slab when its frequency overtakes a resident one, on a
slow background cadence of roughly **one swap per 100 tokens**.

```
current cache : 778 transfers/dispatch = 1.63 GiB/dispatch
tiered slab   : 1 promotion / 100 tokens
              => ~28,000x fewer transfers
```

That is the whole difference between the two designs. The current cache uses the link at
a rate PCIe x4 cannot sustain; this one uses it at a rate where x4 is irrelevant (0.33 ms
of transfer per ~4,700 ms of decode). Tiers converge toward the online policy's measured
46% hit rate **without paying the online policy's traffic**.

## Projected result

| | no CPU/GPU overlap | with overlap |
| --- | --- | --- |
| h=0.41 (measured static heldout prior) | **29.0 t/s (+37%)** | 29.6 (+39%) |
| h=0.46 (after promotion converges) | **30.5 t/s (+44%)** | 31.2 (+47%) |

**The gain does not depend on backend overlap.** If `ggml_backend_sched` serialises the
two passes, +37% still stands. Model: `F=33.3 ms` fixed, 30 routes/layer, 48 layers,
2.78 tokens/dispatch, `t_cpu=0.0677` — fitted on arm 000 and accurate to 4% on configs it
was not fitted to (#130).

Note the design wins at **any** hit rate above zero, because 0.0029 < 0.0677 strictly. A
resident expert is pure profit and a non-resident one costs exactly what arm 000 already
pays. There is no downside case — which is also why the "dead" static prior (heldout
0.406/0.421) is perfectly adequate here: with no transfers, choosing badly is nearly free.

## Work plan

**Step 1 — load-time split and permutation.** `llama-model-loader.cpp`,
`llama-model.cpp`. Build hot/cold slabs and permute `ffn_gate_inp`. Static, no promotion.
*Gate:* greedy-token identity against arm 000, the instrument that passed for #128
(648/648 chars, identical PPL). Byte-identical output or the permutation is wrong.

**Step 2 — two-pass `build_moe_ffn` with null padding.** The +37%.
*Gate:* beats a **same-binary** arm-000 control by >= +25%.

**Step 3 — slow promotion.** Frequency counter plus a swap on a token-count cadence.
*Gate:* h rises toward 0.46; transfer volume stays under 1 promotion/50 tokens.

**Step 4 (optional) — id compaction** to replace null padding. Removes the 3.6% tax.

Steps 1-2 are the result. 3 refines it. 4 is optional.

## Risks

1. **Permutation mismatch between `ffn_gate_inp` and the slabs** routes to wrong experts
   silently, with no crash. Step 1's greedy-identity gate exists for exactly this.
2. **`ffn_gate_inp` is known-fragile in this fork (#128)**: adding a `MUL_MAT` reader of
   it shifts PPL 3.0613 -> 3.1377, deterministically. This design **adds no reader and no
   graph node** — it permutes weights at load — so #128's mechanism should not apply. Verify
   rather than assume.
3. **`ggml_backend_sched` may insert a host/device sync between the two passes**, 96 per
   dispatch. This is the one thing that could undercut the projection; measure it first in
   step 2.
4. **Frequency rankings transfer poorly across workloads** (heldout 0.406/0.421). Priced
   into the table above and the reason step 3 exists.

## Scope

Touches model loading and graph building only. It does **not** touch the plan, reader,
staging or admission code in `moe-cache.cu` — that code is bypassed, not modified.
#124 (cache + MTP mutually exclusive) does not apply, since no cache is active.

## Context

- Design record and full measured history: PR #131 (`docs/moe-expert-tiering-design.md`)
- Measured record of the x4 finding and the five failed arms: #130, #129
- Hardware constraint accepted as given: the 3060 sits on a PCIe x4 root port and will
  stay there.

---

## Step 1 addendum: permute *every* expert-dim tensor, and assert it

Raised in review — the original text named three weight tensors and the router. That is
**not the full set**, and a missed one routes to the wrong expert silently, with no crash.
Enumerated from `src/llama-model.cpp`:

**Weights, expert index on `ne[2]`:**

| tensor | shape | required |
| --- | --- | --- |
| `ffn_gate_exps` | `{n_embd, n_ff, n_expert}` | yes |
| `ffn_up_exps` | `{n_embd, n_ff, n_expert}` | yes |
| `ffn_down_exps` | `{n_ff, n_embd, n_expert}` | yes |
| `ffn_gate_up_exps` | `{n_embd, n_ff*2, n_expert}` | **NOT_REQUIRED** (fused variant) |

**Per-expert scale vectors `{n_expert}` — all `TENSOR_NOT_REQUIRED`:**

`ffn_gate_exps_s`, `ffn_down_exps_s`, `ffn_up_exps_s` (llama-model.cpp:1586/1589/1592)
`ffn_gate_exps_in_s`, `ffn_down_exps_in_s`, `ffn_up_exps_in_s` (:1644/1647/1650)

**Router:** `ffn_gate_inp` (output rows), and `exp_probs_b` `{n_expert}` if present —
it is an `ADD` onto the router logits in the arch table, so it carries the expert index
even though it is not a weight.

Also check the `ffn_*_chexps` family if COMBINED/chunked mode is ever enabled on this
path; `MOE_EXPS_PATTERN` in `src/llama.cpp` matches `(ch|)exps`, so both spellings exist
in this fork.

### Do not hardcode this list

Every one of these is `TENSOR_NOT_REQUIRED`, so presence varies by quantisation and the
list will drift as the arch changes. **Implement it as a sweep plus an assertion, not a
literal list:**

1. Permute every layer tensor that has a dimension equal to `n_expert`.
2. **Assert at end of load that no tensor with an `n_expert`-sized dim was left
   unpermuted.** Abort with the tensor name if one was.

That assertion is the durable guard. It converts this entire failure class — the one that
has silently cost this track five arms — from "hope the list is complete" into a load-time
abort naming the offender. It costs a few lines and it will outlive this issue.

**Comment by ddvnguyen** [comment]
## Colibri cross-check: the design is confirmed, but one gate in this issue is wrong

Researched against the real colibri source at `/mnt/SSD/colibri-1424/`. Three things
matter for #132.

### 1. Colibri uses indirection, never renumbering — confirmed

They keep original expert ids everywhere and resolve placement through a lookup table:

```c
static QSlot *qs(int layer, int eid){ return &G.slot[(size_t)layer*G.ne + eid]; }
```
`c/qwen36_tier.c:62`

```c
int i = lc->slot_by_expert[eid];
if (i < 0 || i >= lc->n || lc->slots[i].eid != eid) return NULL;
```
`c/qwen36.c:674-685` — note the validation: a stale entry degrades to a **miss**, never a
wrong-expert hit.

Renumbering does not appear anywhere in their source. Instead they made placement-only an
**invariant**: "Placement never changes routing or precision"
(`docs/qwen36-cuda-tier.md:30`), "pins decide which bytes are resident, never what the
model computes" (`c/route_trace.h:173-175`), verified as "output bit-identical in all four
runs" (`docs/qwen36-cuda-tier.md:83`).

We learned the same rule by violating it. The no-renumber redesign is the right call.

### 2. THE GATE IN THIS ISSUE IS WRONG — correct it before step 1 runs

`docs/cuda.md:110-118` records that moving experts between CPU and CUDA **changes
floating-point accumulation order**, and that "a current greedy A/B diverged after a near
tie even though both weight paths were byte-exact."

That is a *different* mechanism from the renumbering failure, and it applies to us: our
experts dequantize from iq3_xxs/iq3_s/iq4_nl into float, so a mixed CPU/GPU split cannot
be bit-identical to a pure-CPU control. **The "degenerate splits must be bit-identical"
gate as written would fail for a legitimate reason and could kill a working
implementation.**

Restated gate:

| config | reference | requirement |
| --- | --- | --- |
| `N_hot = 0` (all cold/CPU) | arm 000 (`--cpu-moe`) | **bit-identical** |
| `N_hot = all` (all hot/GPU) | an **all-GPU** control | **bit-identical** |
| any mixed split | arm 000 | **not** bit-identical by construction — gate on PPL within noise plus greedy agreement, accepting near-tie flips |

The two degenerate legs still prove the split machinery is sound, because each keeps every
expert on one device. Only the mixed case admits FP reordering.

### 3. Miss path independently confirmed

Colibri reached our conclusion from the other direction: "Streaming experts deliberately
remain on the original CPU path: copying an expert from NVMe to the GPU on every use would
only replace the disk bottleneck with a PCIe bottleneck" (`docs/cuda.md:3-6`), and the
tier "does not quantize, prune, substitute, or move expert weights over PCIe during
decode" (`docs/cuda.md:71-74`). Decode misses go to CPU compute overlapped with the
in-flight GPU groups (`c/qwen36.c:1982-2029`).

**Their PCIe link width is NOT stated for that rig** — but it does not matter, and that is
the point: the link is off the decode critical path by construction, so their architecture
is link-width-independent. That is exactly why it survives our x4.

### 4. Patterns worth copying for step 3 (promotion)

- Background uploader thread owns all H2D; decode never blocks (`c/qwen36_tier.c:136-206`)
- Bounded queue, **drops when full rather than blocking** (`QT_QCAP 48`, `c/qwen36_tier.c:16`, `:796-818`)
- Promote check every **16 tokens**, heat decay every **1024** (`c/qwen36_tier.c:990-999`)
- Admission hysteresis **25% + 4** (`c/tier.h:9-14`)
- Heat table persisted across runs and reloaded with `>>1` decay (`c/qwen36_tier.c:595-609`, `:1106-1116`)

### 5. One lever we have not considered

They place the **dense trunk** in VRAM by bytes-saved-per-byte economics, charged against
the same expert budget (`docs/qwen36-cuda-tier.md:58-74`, `c/qwen36_tier.c:399-440`). #132
only tiers experts. Worth pricing separately.

### Caveat on their numbers

Their reported gain (0.35 t/s CPU-only → 9.9-11.3 t/s tiered, `docs/qwen36-cuda-tier.md:140-148`,
Threadripper 3945WX + RTX 3070 + Quadro RTX 4000) is measured against a very different
baseline from ours (21.3 t/s). Their hit-rate-to-throughput curve does **not** transfer;
only the architecture does.
